detectron2/modeling/proposal_generator/rpn.py

```diff
@@ -431,24 +433,27 @@ def forward(
             for x in pred_anchor_deltas
         ]
 
-        if self.training:
+        if self.training and not torch.jit.is_scripting():
```

Comment:
I found that we can remove some hacks by doing this (requires https://github.com/pytorch/pytorch/pull/41496):
```diff
--- a/detectron2/modeling/proposal_generator/rpn.py
+++ b/detectron2/modeling/proposal_generator/rpn.py
@@ -266,8 +266,10 @@
         label.scatter_(0, neg_idx, 0)
         return label
 
+    @torch.jit.unused
     @torch.no_grad()
-    def label_and_sample_anchors(self, anchors: List[Boxes], gt_instances: List[Instances]):
+    def label_and_sample_anchors(self, anchors: List[Boxes], gt_instances:
+            List[Instances]) -> Tuple[List[torch.Tensor], List[torch.Tensor]]:
         """
         Args:
             anchors (list[Boxes]): anchors for each feature map.
@@ -323,14 +325,15 @@
             matched_gt_boxes.append(matched_gt_boxes_i)
         return gt_labels, matched_gt_boxes
 
+    @torch.jit.unused
     def losses(
         self,
-        anchors,
+        anchors: List[Boxes],
         pred_objectness_logits: List[torch.Tensor],
         gt_labels: List[torch.Tensor],
         pred_anchor_deltas: List[torch.Tensor],
         gt_boxes: List[torch.Tensor],
-    ):
+    ) -> Dict[str, torch.Tensor]:
         """
         Return the losses from a set of RPN predictions and their associated ground-truth.
 
@@ -390,10 +393,13 @@
             reduction="sum",
         )
         normalizer = self.batch_size_per_image * num_images
-        return {
+        ret = {
             "loss_rpn_cls": objectness_loss / normalizer,
             "loss_rpn_loc": localization_loss / normalizer,
         }
+        losses = {k: v * self.loss_weight.get(k, 1.0) for k, v in
+                ret.items()}
+        return losses
 
     def forward(
         self,
@@ -433,15 +439,12 @@
             for x in pred_anchor_deltas
         ]
 
-        if self.training and not torch.jit.is_scripting():
+        if self.training:
+            assert gt_instances is not None, "RPN requires gt_instances in training!"
             gt_labels, gt_boxes = self.label_and_sample_anchors(anchors, gt_instances)
-            losses = self.losses(
+            return self.losses(
                 anchors, pred_objectness_logits, gt_labels, pred_anchor_deltas, gt_boxes
             )
-            # Currently dict comprehension is not supported by JIT.
-            # https://github.com/pytorch/pytorch/issues/41448
-            for k, v in losses.items():
-                losses[k] = v * self.loss_weight.get(k, 1.0)
         else:
             losses = {}
         proposals = self.predict_proposals(
···

Reply:
looks better! Modification is much cleaner by moving codes to a context that is decorated by `torch.jit.unused`